Adds "Register Domain" CTA if the site has an unused domain credit

tylermarien opened the pull request:
Fixes #9815 Subtask: Create a Call to Action on My Site screen

This PR adds the ability to register a domain if there is an available domain credit. On the **My Site** screen, it will check the current plan for the selected site to see if there is any available domain credit. If there is a domain credit, it will show a call to action which allows the user to redeem that domain credit.

I decided not to show an error message to the user if the plans fail to load. My assumption is that in most cases there is no domain credit, so it would be better to just not show the error if the plans fail to load rather than show an unexpected error message that might confuse the user. I'd be happy to change if this does not make sense.

To test:
1. Install the wasabi build.
2. Switch to a site on a business plan with unclaimed domain credit 
3. The CTA should be displayed near the top of the screen, click on "Register Domain"
4. Pick a domain (preferably .blog).
5. Tap on "CHOOSE DOMAIN" and navigate to contact form details.
6. Fill out the your contact details (if they haven't already been prefilled)
7. Click "REGISTER DOMAIN"
8. If successful, the **Success** screen should be displayed. Click on "NEXT". You should now be returned to the **My SIte** screen

PS: Follow domain testing guideline and make sure if you purchase a domain to cancel it.

![Screenshots](https://user-images.githubusercontent.com/897223/57574107-2022af00-73f0-11e9-9224-d308c703e32a.png)

tylermarien commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java:
```diff
@@ -547,6 +571,48 @@ public void onClick(View v) {
     }
 
+    private void setupObservers(View rootView) {
+        final View domainRegistrationCta = rootView.findViewById(R.id.my_site_register_domain_cta);
```
Good idea. Done.

tylermarien commented on WordPress/src/test/java/org/wordpress/android/viewmodel/main/MySiteViewModelTest.kt:
```diff
@@ -0,0 +1,204 @@
+            viewModel.onPlansFetched(OnPlansFetched(site, plans, null))
+        }
+
+        viewModel.currentPlan.observeForever { plan ->
```
The only reason I used a mock for the siteObserver was because I wanted to assure that it wasn't called in the `setSiteWillNotFireObserversIfNewSiteHasSameIdAsOldSite` test. Using a mock was the only way I could think of to verify that. I try to use real objects in tests when possible because it is possible to test the wrong thing (or nothing at all) when you create a mock that doesn't act like the real object.

tylermarien commented on WordPress/src/main/java/org/wordpress/android/viewmodel/main/MySiteViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+    private val plans = MutableLiveData<List<PlanModel>>()
+    private val status = MutableLiveData<NetworkStatus>()
+
+    val currentPlan = MediatorLiveData<PlanModel>().apply {
```
Good point. I have switched to using `Transformations.map` instead, which will return a LiveData object. I also changed to `isDomainRegistrationVisible` based on your comment above. It allows us to move more logic into the view model.

tylermarien commented on WordPress/src/main/java/org/wordpress/android/viewmodel/main/MySiteViewModel.kt:
```diff
@@ -0,0 +1,82 @@
+import javax.inject.Inject
+
+class MySiteViewModel @Inject constructor(val store: SiteStore, val dispatcher: Dispatcher) : ViewModel() {
+    private val site = MutableLiveData<SiteModel>()
```
Sure. I have generally been following the convention based on what I've seen in the [Android documentation](https://developer.android.com/topic/libraries/architecture/viewmodel#implement) but I like using the property access syntax better.

tylermarien commented:
I have fixed the padding issue. The animation definitely shouldn't run when the orientation changes. We should be able to keep the previous state without loading anything. I will look into that issue and fix.

Regarding your bigger question about the use of a view model. I generally try to use view models and live data whenever possible. It reduces the size of Activity/Fragments (which tend to get bloated). It makes the logic more testable. The use of LiveData also reduces the number of places where state needs to be handled. 

I do agree this is maybe not the ideal implementation. It would probably make more sense if we listened for the OnSiteChanged event inside of the view model, then updated the plans based on that. There would only be one-way data flow from the view model to the fragment. It appears that the OnSiteChanged model doesn't return the new SiteModel though so that is not possible.

My thought when implementing this version of the view model is that it is an incremental step in the right direction. I agree that refactoring the entire Fragment to use a view model would be outside of the scope of this task since it would be a very large change. It might be best to leave the view model refactoring for it's own task as it is likely to get stuck in a half-baked view model state. I do like that we are able to unit test some of this logic though and would hate to lose that.

I can refactor to remove the view model if you think that is the best course of action.

SylvesterWilmott commented:
Looking good @tylermarien !

>Maybe we can add some padding to the call to action description and align it with the rows?

>Also, view animation looks great for the first time, but it keeps happening every time after orientation change. Wonder if we should do something about it?

Agree with both of these. 

- The description text should align with the rows.
- Let's keep the entrance animation for the first time only

tylermarien commented:
Hey @khaykov. I have refactored the PR to remove MySiteViewModel. I extracted a couple of functions (`getCurrentPlan()` and `isDomainCreditAvailable()`) because that logic was duplicated between MySiteFragment and PluginDetailActivity.

The animation is now fixed. It will only display when the site changes. It will not display when the device is rotated.

All of the outstanding comments / issues should be addressed.

![DomainRegistrationCtaRotation](https://user-images.githubusercontent.com/897223/57991113-ec87fa80-7a69-11e9-9d96-4bd55b5c7f38.gif)

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -963,6 +965,18 @@ public void onAccountChanged(OnAccountChanged event) {
+        } else {
+            MySiteFragment fragment = getMySiteFragment();
+            if (fragment != null) {
+                fragment.updateDomainRegistrationCtaIfDomainCreditAvailable(event.plans);
```
`updateDomainRegistrationCtaIfDomainCreditAvailable` changes the visibility of `MySiteFragment` views, but fragment being not null, does not guarantee that its View is created and can be manipulated. In this particular case, `fragment` will never be null (probably)  - if you go deeper into `getMySiteFragment()` you will see that it's lazy-loading `MySiteFragment`. This will cause an unrecoverable crash (you will have to reinstall the app to fix it) if you switch to another tab (from the bottom nav bar) and restart the app.

tylermarien replied:
Great catch. I have addressed this issue in the latest round of changes. I only call `fetchPlansIfNecessary` from `onResume`. It _seems_ to catch all cases where the current site has been changed. I was originally trying to call it in `onSiteChanged` as well, but it creates the same issue since `onSiteChanged` is called using `getMySiteFragment`.

khaykov commented:
Also, `issue/9739-domain-contact-info` was merged into `develop`, so you can update your target branch :)

tylermarien commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java:
```diff
@@ -153,12 +158,14 @@
     private View mQuickStartGrowView;
     private View mQuickStartMenuButton;
     private Handler mQuickStartSnackBarHandler = new Handler();
+    private LinearLayout mDomainRegistrationCta;
```
Good idea. It is now changed.

tylermarien commented:
All changes should be addressed @khaykov. For some reason I thought all network logic was in WPMainActivity. It definitely makes much more sense to contain it all in MySiteFragment. I have refactored to move all the logic in there. It makes a lot more sense now.

I am not sure how I feel about having both `mIsDomainRegistrationCtaVisible` and `mDomainRegistrationCtaSiteId` as member variables. I was gonna create a member named `mCurrentPlan` to store the current plan, from which we could determine both the site id and the visibility. It appears that PlanModel is neither Parcelable or Serializable, so I couldn't save it across screen rotations. Do you have any opinion/suggestions on this?

khaykov commented:
Changes look good! Regarding `PlanModel`, it's part of [Flux-C](https://github.com/wordpress-mobile/WordPress-FluxC-Android), our networking and persistence library. You are free to work with that repo too if you like :)

Wanted to ask, what would happen if I register a domain name through plugin installation flow? When you try to install a plugin without a custom domain, you are prompted to spend your domain credit. You enter the flow from [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/plugins/PluginDetailActivity.java#L311). Looks like we will still the CTA when we return to MySiteFragment after spending credit. What do you think would be a good way to address this?

tylermarien commented:
That's a good question. There are a few options.

1. We could store the state in SharedPreferences. We could store `is_domain_credit_available` in the SharedPreferences. When a domain is registered we could clear that state which would force MySiteFragment to fetch the plans again to calculate the state. This state could be used to decide if the CTA is visible and if we should show the DomainRegistrationDialog.

2. We could store the plans in a local database. We would check the local database to see if we have plans loaded (and that they are related to the current site). If the plans are loaded then we don't have to fetch them again, but if they are not loaded then we would have to fetch them. We would clear out the local database when we register a domain or change a site. This would force MySiteFragment to fetch the plans again from the server. We could also use these plans to decide if the CTA is visible and if we should show the DomainRegistrationDialog.

They both have their positive and negatives. The first option is easier to implement and would require less code changes. The second option is more flexible because we can use if whenever we need to fetch the plans. I personally like the second plan since it does offer that flexibility, but if we don't foresee using the plans in other places then the first plan is easy and quick to implement.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainRegistrationResultFragment.kt:
```diff
@@ -41,6 +43,7 @@ class DomainRegistrationResultFragment : Fragment() {
         checkNotNull((activity?.application as WordPress).component())
 
         continue_button.setOnClickListener {
+            activity!!.setResult(RESULT_OK, Intent())
```
Wonder, why are you passing empty `Intent()` to `setResult()` here?

tylermarien replied:
Hmmmm. Thats a good question. Not sure if I have a good answer. For some reason I thought the second argument on `setResult()` was required. Not sure why I was thinking that. I have removed the empty `Intent()`.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java:
```diff
@@ -573,6 +568,13 @@ public void onClick(View v) {
+    @Override
+    public void onClick(View view) {
+        if (view.getId() == R.id.row_register_domain) {
+            ActivityLauncher.viewDomainRegistrationActivity(this, getSelectedSite());
```
Just to be consistent with the rest of the code, we should probably use inline `onClick` with `R.id.row_register_domain` too :)

tylermarien replied:
Good call. I needed to access the instance of `MySiteFragment` so I added `onClick` to the fragment class. This allowed me to use `this` to reference the fragment instance. I have made it an inline `onClick` as that is much more consistent with the result of the code. Thanks.

nbradbury commented:
Just a friendly reminder to please make sure to manually merge `develop` into this branch before you merge this PR, as `develop` now contains the AndroidX migration. The automatic merge can succeed but still break the build. Better to find out now before it breaks `develop`. Thanks!

khaykov commented:
Closing this PR. We updated to AndroidX so I'm going to bring commits in this PR to the current repo so I can make necessary changes, and continue with the feature.
